### PR TITLE
Make identity work in 1.16.3

### PR DESCRIPTION
**Pull request by AllanChain**
**WIP...**

Updated yarn mapping and switched UI library to [CottonMC/LibGui](https://github.com/CottonMC/LibGui) (because of #40 )

Things seems working now, I want to know your thoughts before proceeding (clean code, enhance UI, implement mouse scrolling, etc.)

PS: I am not a Java programer. So thanks for your review in advance!

### CODE REVIEW

**Comment by Draylar** [comment]
Great start, thanks for your effort into this. My main concern is keeping the UI similar to what it was before, but I'm still down for anything that might improve on the original design. Will look into this more when I get home tonight.

**Comment by AllanChain** [comment]
@Draylar Could you explain why not use LibGui in this PR?

**Comment by Draylar** [comment]
I did not realize Pyro's PR updated the game version to 1.16.3 until after I merged it, then I had to do something else, so I didn't get around to solving this. I still want to merge your PR on top of the CCA changes.

**Comment by Draylar** [comment]
The two main issues I see right now is the scrollbar not working and entities not lighting up when you click on them.

File: src/main/java/draylar/identity/screen/IdentityGui.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import draylar.identity.cca.UnlockedIdentitysComponent;
+import draylar.identity.screen.widget.WEntityButton;
+import draylar.identity.screen.widget.WPlayerButton;
+import draylar.identity.screen.widget.WSearchBar;
+import draylar.identity.registry.Components;
+
+public class IdentityGui extends LightweightGuiDescription {
+    private final List<LivingEntity> renderEntities = new ArrayList<>();
+    private List<WEntityButton> renderedButtons = new ArrayList<>();
+    public IdentityGui() {
```
Comment (Draylar) [inline src/main/java/draylar/identity/screen/IdentityGui.java]: Formatting changes, newline before fields in classes, and newlines splitting methods

```suggestion
public class IdentityGui extends LightweightGuiDescription {

    private final List<LivingEntity> renderEntities = new ArrayList<>();
    private List<WEntityButton> renderedButtons = new ArrayList<>();
    
    public IdentityGui() {
```

File: src/main/java/draylar/identity/screen/IdentityGui.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            List<LivingEntity> filtered = unlocked
+                    .stream()
+                    .filter(livingEntity -> livingEntity.getType().getTranslationKey().contains(widget.getText()))
+                    .collect(Collectors.toList());
+
+            populateEntityWidgets(container, filtered);
+        });
+
+        root.validate(this);
+    }
```
Comment (Draylar) [inline src/main/java/draylar/identity/screen/IdentityGui.java:105]:
```suggestion
    }
    
```

File: src/main/java/draylar/identity/screen/IdentityGui.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            Registry.ENTITY_TYPE.forEach(type -> {
+                Entity entity = type.create(MinecraftClient.getInstance().world);
+
+                // only add living entities to cache
+                if(entity instanceof LivingEntity) {
+                    renderEntities.add((LivingEntity) entity);
+                }
+            });
+        }
+    }
```
Comment (Draylar) [inline src/main/java/draylar/identity/screen/IdentityGui.java:118]:
```suggestion
    }
    
```

File: src/main/java/draylar/identity/screen/IdentityGui.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            renderedButtons.add(button);
+            container.add(
+                button,
+                (getWindow().getScaledWidth() - 27) / 7 * x + 13,
+                (int) Math.ceil(getWindow().getScaledHeight() / 5 * (y + 0.5)),
+                (getWindow().getScaledWidth() - 27) / 7,
+                getWindow().getScaledHeight() / 5
+            );
+        }
+    }
```
Comment (Draylar) [inline src/main/java/draylar/identity/screen/IdentityGui.java:137]:
```suggestion
    }
    
```

File: src/main/java/draylar/identity/screen/IdentityGui.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+        // collect current unlocked identities (or allow all for creative users)
+        renderEntities.forEach(entity -> {
+            if(unlockedIdentitys.has(entity.getType()) || MinecraftClient.getInstance().player.isCreative()) {
+                unlocked.add(entity);
+            }
+        });
+
+        return unlocked;
+    }
```
Comment (Draylar) [inline src/main/java/draylar/identity/screen/IdentityGui.java:150]:
```suggestion
    }
    
```

File: src/main/java/draylar/identity/screen/IdentityScreen.java [hunk 6]
```diff
@@ -1,172 +1,21 @@
...
-//                            .setPosition(Position.of(container, (getWindow().getScaledWidth() - 27) / 7f * x, getWindow().getScaledHeight() / 5f * y)));
-
-                    widgets.add(button);
-                }
-            }
-
-            container.addRow(widgets.toArray(new WAbstractWidget[widgets.size()]));
-        }
+        super(new IdentityGui());
     }
```
Comment (Draylar) [inline src/main/java/draylar/identity/screen/IdentityScreen.java:18]:
```suggestion
    }
    
```

File: src/main/java/draylar/identity/screen/IdentityGui.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        // get identity components from player
+        UnlockedIdentitysComponent unlockedIdentitys = Components.UNLOCKED_IDENTITIES.get(MinecraftClient.getInstance().player);
+        IdentityComponent currentIdentity = Components.CURRENT_IDENTITY.get(MinecraftClient.getInstance().player);
+
+        // collect unlocked entities
+        List<LivingEntity> unlocked = collectUnlockedEntities(unlockedIdentitys);
+
+        populateEntityWidgets(container, unlocked);
+
+        WPlayerButton playerButton = new WPlayerButton();
```
Comment (Draylar) [inline src/main/java/draylar/identity/screen/IdentityGui.java:68]: I would still like to keep the hint that was around here. Was there any reason for removing it?

**Review by Draylar — CHANGES_REQUESTED** [review]

File: src/main/java/draylar/identity/screen/IdentityGui.java (hunk 7 above)
Comment (AllanChain) [reply src/main/java/draylar/identity/screen/IdentityGui.java:68]: Oh sorry, I removed some comments by accident 😢

**Comment by Draylar** [comment]
Got Spinnery working on 1.16.2+ because I didn't like the LibGUI scrollbar.

image

At this point I'm thinking maintaining the Spinnery fork is the best route, unless you have any ideas on fixing the GUI buttons and scrolling.

**Comment by AllanChain** [comment]
It's true that scrollbar in spinnery is better. I think it can be fixed but hundreds of lines of code needs to be added.

As you have other mods depending on spinnery, I think maintaining a fork is a good choice :)

**Comment by James103** [comment]
Is this pull request still relevant? If not, this can be closed.
Identity already works on 1.16.3 if that was the original intention.

**Comment by AllanChain** [comment]
This PR is stale and outdated. And the forked Spinnery works fine. I am closing this.

ps: LibGui has implemented better scrollbar (CottonMC/LibGui/pull/90, CottonMC/LibGui/pull/91), But without momentum scrolling. LibGui may still be a good option if maintaining Spinnery fork becomes a pain in the future.

**Pull request closed without merging**